### PR TITLE
Function node

### PR DESCRIPTION
**Pull request by shanemadden**
Here's my current draft on the function node, zwave-cmd, to have a way to make calls to the OZW API and get the return values back into messages.

It's not quite to where I want it to be, particularly, I want to go through the list of function calls in the options dropdown and add their arguments to the dropdown string, for reference - but I wanted to get some eyes on this in the current state to make sure it looks like a good direction to spend more time on.

This node, as the experimental API support on the zwave-out node, uses the topic as the function name (I actually just made a couple modifications to the existing node's function to make it work for both node types).  I'm not sure if we should necessarily keep using the topic or if moving to something like message.command would be more proper node-red behavior, but I don't mind shuffling stuff around.

There's a topic dropdown to allow the node to set the function call, which includes all of the function calls from https://github.com/OpenZWave/node-openzwave-shared/blob/master/src/openzwave.cc at the moment (without having tested that they all work, yet).

Let me know how this looks, I'll spend some more time on the topic dropdown if this seems like a good direction.

### CODE REVIEW

**Comment by ekarak** [comment]
are you sure this works? What I mean is you're not defining a new node, you're merely adding an alias to the existing zwave-out node... I'll comment on the code.

File: 10-zwave.js [hunk 1]
```diff
@@ -410,7 +410,16 @@ module.exports = function(RED) {
...
+							var retval = ozwDriver[msg.topic].apply(ozwDriver, payload);
+							var newmsg = {'topic': msg.topic};
+							if (retval) {
+								if (debug) console.log('Got return value, sending as payload');
+								newmsg.payload = retval;
+							} else {
+								if (debug) console.log('No return value but call successful, sending empty array');
+								newmsg.payload = [];
+							}
+							node.send(newmsg);
```
Comment (ekarak) [inline 10-zwave.js]: Wont this addition cause the zwave-out node to send out messages to NR flows as well?

File: 10-zwave.js [hunk 2]
```diff
@@ -410,7 +410,16 @@ module.exports = function(RED) {
 					) {
 						if (debug) console.log('attempting direct call to OpenZWave API: %s(%s)', msg.topic, payload);
 						try {
-							ozwDriver[msg.topic].apply(ozwDriver, payload);
+							var retval = ozwDriver[msg.topic].apply(ozwDriver, payload);
+							var newmsg = {'topic': msg.topic};
```
Comment (ekarak) [inline 10-zwave.js]: What about *async* calls to OpenZWave? You won't be getting a return vallue from the call here, but through a callback some milliseconds later (thus: from the zwave-in node)

File: 10-zwave.js (hunk 1 above)
Comment (shanemadden) [inline 10-zwave.js]: It would attempt to send, but effectively do nothing since there's no output wires on the zwave-out node. Tweaking the zwave-out node's behavior was the lowest impact way to just glue the functionality on without stripping experimental API support from the zwave-out node (yet).

I feel like long term, we'd want to split them out - with the new node type handling all the direct API stuff, zwave-out handing what it's always handled, and with the them both having different constructors.

File: 10-zwave.js (hunk 2 above)
Comment (shanemadden) [inline 10-zwave.js]: Currently, it's just completely naive about async calls, and the behavior is as you're describing (aka, kind of disjointed and unintuitive), the results of async calls will pile in a zwave-in node, and the user doesn't have a way to know what was in response to what call.  It can be made to work with all your data going to the right place, but it's not super pretty; it's most useful for synchronous calls at the moment.

I haven't fully wrapped my head around how I'd build a "wait for the callback, then emit returns from that event" mechanism, but at the very least we'd need to refactor event handling to "catch" the callbacks selectively that originated from async calls, have the code somehow know whether to expect a callback for a given call and what that callback would look like, and have a tracking mechanism for every pending callback stored in context (which may take milliseconds or never return at all, if a node never wakes up).

In other words, that seemed like a little too much for me to bite off all at once, but it'll be awesome if we can get there.

**Comment by ekarak** [comment]
One other thing to worry about is the version of OpenZWave library. All the management functions (eg adding a new node) have been (up to version 1.3 IIRC) incorporated into one generic `BeginControllerCommand` call, and now this generic call has been deprecated in favour of the individual calls (eg. `AddNode`). We _could_ in theory require a recent OpenZWave, but this will cause grief to some users.

**Comment by shanemadden** [comment]
That's a good point, hmm.. I'm not sure if it's worth building glue code for the older OpenZWave, or just maybe have the new node type just not register itself if the underlying OpenZWave doesn't support the calls it would be making?
